File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteTaskHierarchyRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.firebase.DatabaseWrapper;
import com.krystianwsul.checkme.firebase.json.TaskHierarchyJson;

import junit.framework.Assert;

public class RemoteTaskHierarchyRecord extends RemoteRecord {
    public static final String TASK_HIERARCHIES = "taskHierarchies";

    @NonNull
    private final String mId;

    @NonNull
    private final RemoteProjectRecord mRemoteProjectRecord;

    @NonNull
    private final TaskHierarchyJson mTaskHierarchyJson;

    RemoteTaskHierarchyRecord(@NonNull String id, @NonNull RemoteProjectRecord remoteProjectRecord, @NonNull TaskHierarchyJson taskHierarchyJson) {
        super(false);

        mId = id;
        mRemoteProjectRecord = remoteProjectRecord;
        mTaskHierarchyJson = taskHierarchyJson;
    }

    RemoteTaskHierarchyRecord(@NonNull RemoteProjectRecord remoteProjectRecord, @NonNull TaskHierarchyJson taskHierarchyJson) {
        super(true);

        mId = DatabaseWrapper.INSTANCE.getTaskHierarchyRecordId(remoteProjectRecord.getId());
        mRemoteProjectRecord = remoteProjectRecord;
        mTaskHierarchyJson = taskHierarchyJson;
    }

    @NonNull
    @Override
    public TaskHierarchyJson getCreateObject() {
        return mTaskHierarchyJson;
    }

    @NonNull
    public String getId() {
        return mId;
    }

    @NonNull
    @Override
    protected String getKey() {
        return mRemoteProjectRecord.getKey() + "/" + RemoteProjectRecord.PROJECT_JSON + "/" + TASK_HIERARCHIES + "/" + mId;
    }

    public long getStartTime() {
        return mTaskHierarchyJson.getStartTime();
    }

    @Nullable
    public Long getEndTime() {
        return mTaskHierarchyJson.getEndTime();
    }

    @NonNull
    public String getParentTaskId() {
        return mTaskHierarchyJson.getParentTaskId();
    }

    @NonNull
    public String getChildTaskId() {
        return mTaskHierarchyJson.getChildTaskId();
    }

    public void setEndTime(long endTime) {
        Assert.assertTrue(getEndTime() == null);

        mTaskHierarchyJson.setEndTime(endTime);
        addValue(getKey() + "/endTime", endTime);
    }

    @Nullable
    public Double getOrdinal() {
        return mTaskHierarchyJson.getOrdinal();
    }

    public void setOrdinal(double ordinal) {
        mTaskHierarchyJson.setOrdinal(ordinal);
        addValue(getKey() + "/ordinal", ordinal);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteTaskHierarchyRecord.kt
package com.krystianwsul.checkme.firebase.records

import com.krystianwsul.checkme.firebase.DatabaseWrapper
import com.krystianwsul.checkme.firebase.json.TaskHierarchyJson

import junit.framework.Assert

class RemoteTaskHierarchyRecord : RemoteRecord {

    companion object {

        const val TASK_HIERARCHIES = "taskHierarchies"
    }

    val id: String

    private val remoteProjectRecord: RemoteProjectRecord

    override val createObject: TaskHierarchyJson

    override val key get() = remoteProjectRecord.key + "/" + RemoteProjectRecord.PROJECT_JSON + "/" + TASK_HIERARCHIES + "/" + id

    val startTime get() = createObject.startTime

    val endTime get() = createObject.endTime

    val parentTaskId get() = createObject.parentTaskId

    val childTaskId get() = createObject.childTaskId

    val ordinal get() = createObject.ordinal

    constructor(id: String, remoteProjectRecord: RemoteProjectRecord, taskHierarchyJson: TaskHierarchyJson) : super(false) {
        this.id = id
        this.remoteProjectRecord = remoteProjectRecord
        createObject = taskHierarchyJson
    }

    constructor(remoteProjectRecord: RemoteProjectRecord, taskHierarchyJson: TaskHierarchyJson) : super(true) {
        id = DatabaseWrapper.getTaskHierarchyRecordId(remoteProjectRecord.id)
        this.remoteProjectRecord = remoteProjectRecord
        createObject = taskHierarchyJson
    }

    fun setEndTime(endTime: Long) {
        Assert.assertTrue(createObject.endTime == null)

        createObject.endTime = endTime
        addValue("$key/endTime", endTime)
    }

    fun setOrdinal(ordinal: Double) {
        createObject.ordinal = ordinal
        addValue("$key/ordinal", ordinal)
    }
}
